### ISSUE TITLE
Endless loop (100% CPU) with shlex.split completion

### ISSUE THREAD

**Issue by blueyed** [davidhalter/jedi#929]
The following script will cause 100% CPU:

```python
import jedi

script = """import shlex
qsplit = shlex.split("foo, ferwerwerw werw werw e")
for part in qsplit:
    part."""

script = jedi.Script(script, 4, 9)
print(script.completions())
```

Via https://github.com/davidhalter/jedi-vim/issues/679#issuecomment-294364150.

Tested it on 0.10.2, master and dev.

**Comment by blueyed** [comment]
Added some printing to `_eval_element_not_cached`, and this is where it starts looping:
```
…
(' self.punctuation_chars', 'atom_expr', PythonNode(atom_expr, [<Name: self@133,15>, PythonNode(trailer, [<Operator: . start=(133, 19)>, <Name: punctuation_chars@133,20>])]), <Name: self@133,15>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.instream.read(1)', 'atom_expr', PythonNode(atom_expr, [<Name: self@136,27>, PythonNode(trailer, [<Operator: . start=(136, 31)>, <Name: instream@136,32>]), PythonNode(trailer, [<Operator: . start=(136, 40)>, <Name: read@136,41>]), PythonNode(trailer, [<Operator: ( start=(136, 45)>, <Number: 1 start=(136, 46)>, <Operator: ) start=(136, 47)>])]), <Name: self@136,27>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
('self.posix', 'atom_expr', PythonNode(atom_expr, [<Name: self@195,22>, PythonNode(trailer, [<Operator: . start=(195, 26)>, <Name: posix@195,27>])]), <Name: self@195,22>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,17>, PythonNode(trailer, [<Operator: . start=(181, 21)>, <Name: state@181,22>])]), <Name: self@181,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@145,17>, PythonNode(trailer, [<Operator: . start=(145, 21)>, <Name: state@145,22>])]), <Name: self@145,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@142,15>, PythonNode(trailer, [<Operator: . start=(142, 19)>, <Name: state@142,20>])]), <Name: self@142,15>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.quotes', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,31>, PythonNode(trailer, [<Operator: . start=(181, 35)>, <Name: quotes@181,36>])]), <Name: self@181,31>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,17>, PythonNode(trailer, [<Operator: . start=(181, 21)>, <Name: state@181,22>])]), <Name: self@181,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@145,17>, PythonNode(trailer, [<Operator: . start=(145, 21)>, <Name: state@145,22>])]), <Name: self@145,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@142,15>, PythonNode(trailer, [<Operator: . start=(142, 19)>, <Name: state@142,20>])]), <Name: self@142,15>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.quotes', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,31>, PythonNode(trailer, [<Operator: . start=(181, 35)>, <Name: quotes@181,36>])]), <Name: self@181,31>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,17>, PythonNode(trailer, [<Operator: . start=(181, 21)>, <Name: state@181,22>])]), <Name: self@181,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@145,17>, PythonNode(trailer, [<Operator: . start=(145, 21)>, <Name: state@145,22>])]), <Name: self@145,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@142,15>, PythonNode(trailer, [<Operator: . start=(142, 19)>, <Name: state@142,20>])]), <Name: self@142,15>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.quotes', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,31>, PythonNode(trailer, [<Operator: . start=(181, 35)>, <Name: quotes@181,36>])]), <Name: self@181,31>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,17>, PythonNode(trailer, [<Operator: . start=(181, 21)>, <Name: state@181,22>])]), <Name: self@181,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@145,17>, PythonNode(trailer, [<Operator: . start=(145, 21)>, <Name: state@145,22>])]), <Name: self@145,17>)
(' self.state', 'atom_expr', PythonNode(atom_expr, [<Name: self@142,15>, PythonNode(trailer, [<Operator: . start=(142, 19)>, <Name: state@142,20>])]), <Name: self@142,15>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
(' self.quotes', 'atom_expr', PythonNode(atom_expr, [<Name: self@181,31>, PythonNode(trailer, [<Operator: . start=(181, 35)>, <Name: quotes@181,36>])]), <Name: self@181,31>)
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
{<TreeInstance of <ClassContext: <Class: shlex@19-299>>(<TreeArguments: PythonNode(arglist, [<Name: s@301,16>, <Operator: , start=(301, 17)>, PythonNode(argument, [<Name: posix@301,19>, <Operator: = start=(301, 24)>, <Name: posix@301,25>])])>)>}
…
```

Relevant shlex code reference: https://github.com/python/cpython/blob/ba6bfcdfb43f06f025613348bb6c71ff9b597b35/Lib/shlex.py#L181.

**Comment by davidhalter** [comment]
That is pretty much the worst code that you can write for Jedi. This is so annoying for backtracing. I'll investigate. Hope it doesn't take too long. It's just an extremely annoying problem.

**Comment by davidhalter** [comment]
To check where this is coming from I added a memoize method for name resolution. This led to massively faster speeds (it still took 30 seconds). But I then realized that the actual issue is flow evaluation (if checks). It did 152'651 checks lol. I then removed the memoization again and checked how many flows where then evaluated and it was even worse. I stopped the program after a few hours and 50'000'000 flow checks. I still think that it wasn't in an infinite loop, but it might have taken days if not years.

Well, I guess I just have to add an upper limit count of flow analysis checks. That's probably the most reasonable thing.

**Comment by blueyed** [comment]
Having (a specialized) memoize there might be good in general then, too?

**Comment by davidhalter** [comment]
Fixed in 8fcb4685390b3b76b168a0633ed59f54e7b29bfd.

I decided to limit type inference per syntax tree scope. I've tried a lot of different things and in the end that seemed to be the only feasible thing.

> Having (a specialized) memoize there might be good in general then, too?

I don't really think that that would help with anything, because basically the evaluation of `if ...:` would turn the memoization off, which means that exactly in this case it wouldn't help. Memoization in jedi is extremely complicated because for complex type inference multiple paths through the same syntax nodes need to be taken.

What I did above would probably result in a lot of failing tests. I was just trying to narrow the issue down.

**Comment by davidhalter** [comment]
Fixed in 8fcb4685390b3b76b168a0633ed59f54e7b29bfd.

I decided to limit type inference per syntax tree scope. I've tried a lot of different things and in the end that seemed to be the only feasible thing.

> Having (a specialized) memoize there might be good in general then, too?

I don't really think that that would help with anything, because basically the evaluation of `if ...:` would turn the memoization off, which means that exactly in this case it wouldn't help. Memoization in jedi is extremely complicated because for complex type inference multiple paths through the same syntax nodes need to be taken.

What I did above would probably result in a lot of failing tests. I was just trying to narrow the issue down.

Thanks for this case, extremely good that we've finally fixed this. This sort of thing is extremely annoying for users.
